Refactor legacy_install to not take an install_req

pfmoore opened the pull request.

pfmoore commented:
Two questions:

* Do refactorings like this need a news entry or is it OK to mark them as trivial?
* We get the global and install options from the requirement and add them to the supplied options. But we don't do that for editable installs - is there any reason for this, or is it just an omission? It would be a functional change rather than a refactoring, so I'd do it in a follow-up PR if we were going to, but it's something to consider.

pradyunsg commented:
> Do refactorings like this need a news entry or is it OK to mark them as trivial?

Trivial; since these aren't use facing changes.

pradyunsg commented:
Nice round number for the PR @pfmoore. :)

pradyunsg commented:
Oooooo... Nice catch @chrahunt!

pfmoore commented:
Thanks @chrahunt - I wasn't entirely convinced about how `install_succeeded` with the wider code, so thanks for spotting this! I'll take a look today.

chrahunt commented:
I'm OK with changing the behavior if it makes sense to do it. I do not think what we do currently was necessarily intentional, and we shouldn't feel obliged to bend over backwards to preserve edge cases like this.

pfmoore commented:
Agreed, but I'd prefer to clean up deliberately rather than accidentally 🙂

pfmoore commented:
The problem is that the rollback just rolls back the uninstall in the case of `--force-reinstall`. That's fine if the installation has failed, but once `setup.py` has run, reverting the uninstall will leave behind files and directories from the new install that weren't present in the old version (crucially, that might include a `.egg-info` directory as it has the project version in the name).

The old code wouldn't ever roll back once `setup.py` had run (technically, once the deprecation period expires, the exception if there was no `.egg-info` directory would roll back, but that's low-impact). The new code could roll back on a subsequent exception, potentially leading to duplicate `.egg-info` directories.

So this does need fixing - it's messy but a necessary consequence of the fact that the legacy path lets `setup.py` put stuff directly into the target location outside of our control.

Nice catch @chrahunt! I'd never have spotted that one.

(Looking forward to the day when we can just remove the legacy install path - or at least the part of it that doesn't build wheels before installing 🙏).

pfmoore commented:
Sigh. Short of splitting the function in two, I think I need to wrap and re-raise any exception in the actual install to fix this. I know how to do this in Python 3 (chained exceptions) but I've no idea how to handle it in a Python 2 compatible way. Does anyone have any idea?

Basically what I'd do in Python 3 is:

```python
    try:
        with indent_log(), build_env:
            runner(
                cmd=install_args,
                cwd=unpacked_source_directory,
            )
    except Exception as exc:
        raise SomeInstallationError from exc
```

and in the caller:

```python
try:
    self.install_succeeded = install_legacy(...)
except SomeInstallationError as exc:
    self.install_succeeded = False
    raise exc.__cause__ from None
```

but that's not even valid Python 2 syntax, and I've no idea what I could do instead. (Probably something horrid with `sys.exc_info()`...) Does anyone have any suggestions? I could just skip the exception chaining, but that would lose any exception context in the traceback. Is that acceptable, for the sake of Python 2?

pradyunsg commented:
@pfmoore I think you can use six's helpers for exception handling.

https://six.readthedocs.io/#six.raise_from

pfmoore commented:
How horrible do people find using a mutable parameter to the function?

```
def install(...,
    install_succeeded,  #type: List[bool]
):
    install_succeeded[0] = False
    # do the install
    install_succeeded[0] = True
```

and in the caller:

```python
    install_succeeded = [False]
    try:
        legacy_install(..., install_succeeded)
    finally:
        self.install_succeeded = install_succeeded[0]
```

I promise I'll add a comment to the code apologising for the existence of Python 2 and what it made me do... ☹️ 

(P.S. I don't want to even think about how I'd write a test for this edge case. I assume everyone is OK with leaving it untested, as it's on the legacy path and it hasn't had a test until now? Or at least, deferring writing a test for a follow-up PR?)

pradyunsg commented:
> How horrible do people find using a mutable parameter to the function?

Very horrible.

I'd much rather lose context in our exceptions on Python 2, rather than have such a hack that needs a comment on how Python 2 still exists. 🙈

pfmoore commented:
OK. So the following seems to work, albeit with some extra clutter on the Python 3 traceback (as well as losing context in Python 2):

```python
class C(Exception): pass

def foo():
    try:
        bar()
    except Exception as exc:
        if isinstance(exc, C):
            print("Got C")
        else:
            print("Did not get C")
        raise

def bar():
    try:
        1/0
    except Exception as exc:
        raise C

foo()
```

I'll go with that, I guess (and hate the need to support Python 2 just a little bit more...)

pfmoore commented:
Doh, that doesn't work because it loses the exception type. I need to stop and think harder.

I the suggestion about six, but when I tried it I hit other difficulties related to the `__cause__` attribute of exceptions. I'll research it a bit further.

chrahunt commented:
Just to ensure forward progress, we could break all the argument changes except removing the install_req into a separate PR.

pfmoore commented:
@chrahunt I think the latest commit fixes the problems here, so I don't think we need to split this up.

pfmoore commented:
@chrahunt I think the latest commit fixes the problems here, so I don't think we need to split this up.

pfmoore commented:
@chrahunt gentle ping. I think this is now good to go - do you have any further reservations?

xavfernandez commented on src/pip/_internal/req/req_install.py:
```diff
@@ -815,14 +816,39 @@ def install(
+                req_description=str(self.req),
+            )
+        except Exception as exc:
+            if isinstance(exc, LegacyInstallFailure):
```
Why not:
```
except LegacyInstallFailure as exc:
    self.install_succeeded = False
    six.reraise(*exc.parent)
except Exception as exc:
    self.install_succeeded = True
    raise
```
?

pfmoore replied:
Hmm. I think that's probably a holdover from a previous version of that construct. Good catch, I'll change it.

pfmoore commented:
I'm assuming the test failures are because of the new virtualenv release and I'll need to rebase on master? Not sure why I didn't get these on my local test run 🙁 

I'll rebase, but it might not be this evening.

pradyunsg commented:
There's no fix in master yet. :(

pradyunsg commented:
ALA #7717. In case this run doesn't work out, anyone @pypa/pip-committers who wants to push things to that branch, is welcome to do so (after making a comment before pushing, so that folks can know what is happening).

pfmoore commented:
> There's no fix in master yet.

Yeah, I realised later that my local tests probably worked because I didn't recreate the tox environments...

pradyunsg commented:
#7717 worked. You can rebase.

pfmoore commented:
Nice, thanks. Will do that this evening.

pradyunsg commented on src/pip/_internal/req/req_install.py:
```diff
@@ -824,14 +825,38 @@ def install(
-            pycompile=pycompile,
-            scheme=scheme,
-        )
+        # TODO: Why don't we do this for editable installs?
```
Why do we do this for non-editable installs? ;)

pfmoore replied:
I have no idea :-( It was in the existing code so I just moved it as part of the refactoring. The TODO was to note that it's odd that we do different things for the 2 cases, and now that the code got moved, that difference is more obvious. But fixing it should be a separate PR.

pradyunsg commented:
@pfmoore You merged without approvals! Are you adopting my bad habits? :P

pfmoore commented:
Um, I'd addressed the points raised in the 2 reviews, and hadn't heard back from @chrahunt after a couple of pings, so I assumed all was good. Is that not OK?

As far as I was aware, we don't have a policy of requiring approvals, although I usually would. This has been around for a while though and was only waiting on the CI to be fixed for the last couple of days, so I've probably had all the reviews I'm getting here...

I'm happy to have a discussion on what policy we should adopt if people want, although I'd argue against a strict policy of needing approvals as we're a small enough group that IMO we'd be creating bottlenecks by doing so.

pradyunsg commented:
@pfmoore Yea, I wasn't suggesting that you shouldn't have merged.

I was joking, but I guess that subtly wasn't clear in my comment.

I don't think we need a policy change here. :)

pfmoore commented:
No problem 🙂 I did get that you were joking, but took the chance to comment on my thoughts on needing approvals. Sorry for being over-pedantic in my reply.

chrahunt commented:
This looks confusing, but less than it did before this PR. :) Thanks @pfmoore and sorry for the delay getting back to you.
